### PR TITLE
Fixed Typescript issues, added gameType selection

### PR DESCRIPTION
**Pull request by Mathuran**
Added a new start page to select what game play you want

image

### CODE REVIEW

File: src/lib/AIPlayer.ts [hunk 1]
```diff
@@ -90,17 +102,18 @@ export type DifficultyType = typeof Difficulty[keyof typeof Difficulty];
  *
  * @return     Best move for selected difficulty
  */
-export function selectMove(game: Game, difficulty: Difficulty = Difficulty.Medium): number {
+export function selectMove(game: Game, difficulty: DifficultyType = Difficulty.Medium): number | undefined {
```
Comment (Khufu-I) [inline src/lib/AIPlayer.ts:105]: `selectMove` should always return a move. If the game is over then the next line should assert and throw an exception.

File: src/lib/AIPlayer.ts [hunk 2]
```diff
@@ -90,17 +102,18 @@ export type DifficultyType = typeof Difficulty[keyof typeof Difficulty];
...
 	console.assert(!game.isGameOver, `Game is over, status ${game.status}`);
 	const moves = game.moves;
 	if (moves.length === 1) return moves[0];
 
 	let maxRating = Number.NEGATIVE_INFINITY;
 	let bestMove = moves[0];
 	for (const move of moves) {
+		if (move == undefined) break;
```
Comment (Khufu-I) [inline src/lib/AIPlayer.ts:113]: `move` should never be undefined this should be typed properly.

File: src/lib/Game.ts [hunk 3]
```diff
@@ -12,11 +12,20 @@ export const Status = {
...
 } as const;
 
 export type GameStatus = typeof Status[keyof typeof Status];
 
+export const GameTypes = {
+	PlayAI: 0,
+	PassAndPlay: 1
+}
+
+export type GameType = typeof GameTypes[keyof typeof GameTypes]
```
Comment (Khufu-I) [inline src/lib/Game.ts:26]: `Game.ts` could be thought as the game engine for Tic Tac Toe and `page.svelte` as the game controller. The game engine doesn't actually care who executes the move whether it's a real player or the AI, all it does is manage the game state according to the rules.
There shouldn't be any changes required to this file to support game type selection. See #16 for an example how the AI and Game engine were integrated on the Kotlin side.

File: src/lib/Game.ts [hunk 4]
```diff
@@ -66,6 +82,14 @@ export default class Game {
 					.filter((elem) => elem !== undefined);
 	}
 
+	startGame(): void {
+		this.#gameStatus = Status.XTurn
+	}
+
+	newGame(): void {
+		this.#gameStatus = Status.SelectGame
+	}
```
Comment (Khufu-I) [inline src/lib/Game.ts:91]: Should not be necessary

File: src/routes/TicTacToe/+page.svelte [hunk 5]
```diff
@@ -3,21 +3,31 @@
 	import 'papercss/dist/paper.min.css';
 
 	import { Button } from 'spaper';
-	import Game, { Status } from '../../lib/Game';
-	import { selectMove } from '../../lib/AIPlayer';
+	import Game, { GameTypes, Status } from '../../lib/Game';
+	import { selectMove as aiPlayerSelectMove } from '../../lib/AIPlayer';
```
Comment (Khufu-I) [inline src/routes/TicTacToe/+page.svelte:7]: You can import as AIPlayer and then call `AIPlayer.selectMove`

```suggestion
	import AIPlayer  from '../../lib/AIPlayer';
```

**Review by Khufu-I — CHANGES_REQUESTED** [review]
If possible could you split Typescript issues from the rest of the code changes?

**Review by Vithop — CHANGES_REQUESTED** [review]
address siyamans concerns

File: src/lib/AIPlayer.ts [hunk 6]
```diff
@@ -56,10 +55,10 @@ function alphabeta(
 	const isSamePlayer = (game.status == Status.XTurn) == isMaximizingPlayerX;
 	let result = isSamePlayer ? Number.NEGATIVE_INFINITY : Number.POSITIVE_INFINITY;
 	for (const move of moves) {
-		if (newAlpha >= newBeta) break;
+		if (newAlpha >= newBeta || move == undefined) break;
 		const nextGame = game.deepCopy();
 		nextGame.executeMove(move);
-		const rating = alphabeta(nextGame, isMaximizingPlayerX, depth - 1, newAlpha, newBeta);
+		const rating = alphabeta(nextGame, isMaximizingPlayerX, lowerDifficulty(depth), newAlpha, newBeta);
```
[the remark this answers is not in the document]
Comment (Mathuran) [reply src/lib/AIPlayer.ts:61]: would changing the name to ``lowerSearchDepth()``  be better

File: src/lib/AIPlayer.ts [hunk 7]
```diff
@@ -81,6 +80,19 @@ export const Difficulty = {
...
+			return 2;
+		case 3:
+			return 2;
+		case 5:
+			return 3;
+		default:
+			return 2
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (Mathuran) [reply src/lib/AIPlayer.ts:95]: yea the AI worked.

[the remark this answers is not in the document]
Comment (Khufu-I) [reply src/lib/AIPlayer.ts:95]: I don’t think you can beat the easy AI with this change. Now  that testing is enabled I can add tests for all this.

File: src/lib/AIPlayer.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Khufu-I) [reply src/lib/AIPlayer.ts:61]: Sure, you only need to decrease by one every iteration. I don’t understand what `lowerDifficulty` is trying to do. The existing code is doing a [depth first search](https://en.m.wikipedia.org/wiki/Depth-first_search)

[the remark this answers is not in the document]
Comment (Mathuran) [reply src/lib/AIPlayer.ts:61]: Oh okay. I didn't understand what the code was doing either cause you were using the wrong types. I should just use a new var with the type Integer, not Difficulty which starts at the difficulty level selected.

[the remark this answers is not in the document]
Comment (Khufu-I) [reply src/lib/AIPlayer.ts:61]: The number in the Difficulty enum is supposed to represent the depth it’s searching the game tree. This should probably be documented better. The [Kotlin code](https://github.com/Vithop/TicTacToe/blob/master/android/app/src/main/java/com/vithop/tictactoe/AIPlayer.kt#L77) is probably easier to understand. I made a mistake and `depth: DifficultyType` should be `depth: number`

@Vithop didn’t setup the Typescript properly and it’s not being checked correctly, so it’s their fault

**Pull request closed without merging**